### PR TITLE
feat: execute projects from within workspace, deprecate shuttle-static-folder, make persist persistent, more metadata

### LINKED ISSUE #808: [Feature]: allow deploying other artifacts alongside executable during build
### Describe the feature

This should be possible so stuff needed at runtime, e.g. a binary-only dynamic library from a commercial vendor or some other runtime dependency that is not part of the build can be downloaded and moved there.

* One way would be to expose the final executable destination during `build.rs`. E.e. via `SHUTTLE_DEPLOY_DIR`, so one could copy stuff there.

* Alternatively this could be done the other way around. I.e. `build.rs` would set an environment variable that contains a list of artifacts that shuttle needs to make sure are available in the same folder as the executable being built.

  For example, if a the crate `foo`'s `build.rs` downloads to `OUT_DIR` and the resp. crate is a dependency of `bar` which is depolyed to shuttle, the artifacts downloads would end up deep in `bar`'s `target` folder:

  ```
  target/debug/build/foo-6039619eac329ac4/artifact
  target/debug/build/foo-6039619eac329ac4/other_artifact
  ```

  So one option would be to set `SHUTTLE_DEPLOY_ARTIFACTS` to 
  ```
  target/debug/build/foo-6039619eac329ac4/artifact:target/debug/build/foo-6039619eac329ac4/other_artifact
  ```
  And shuttle would make sure those are copied from there to be available alongside the executable at runtime.

* Lastly there could be a crate from shuttle that allows just doing this via calling some methods during `build.rs`.

All that matters is that there is a way to deploy stuff beside the executable in the final destination.

### Duplicate declaration

- [X] I have searched the issues and this feature has not been requested before.

### LINKED ISSUE #928: [Bug]: .cargo/config.toml is not deployed
### What happened?

While working on an example app, I found out that `.cargo/config.toml` is not deployed. Yet this is needed to successfully compile, because the app is using uuid v7 which requires setting specific rustflags and that by default happens in `.cargo/config.toml`.

More details can be found at https://discord.com/channels/803236282088161321/1109452835214590123.

To reproduce the issue, use tag shuttle-issue-928: https://github.com/hseeberger/ticklist/releases/tag/shuttle-issue-928.

### Version

0.16.0

### Which operating systems are you seeing the problem on?

macOS

### Which CPU architectures are you seeing the problem on?

ARM64

### Relevant log output

_No response_

### Duplicate declaration

- [X] I have searched the issues and there are none like this.

### LINKED ISSUE #1008: [Bug]: Static folder error in workspace projects local runs
### What happened?

Using a static folder in a project that is a workspace does not work for local runs and deployments for the same location of the static directory, see discord thread: https://discord.com/channels/803236282088161321/1097033679844429844

## Requirements
- A shuttle project that is a workspace should be able to use a static folder for both local runs and deployments, without having to move the static folder to different locations for each environment.

### Version

v0.18.0

### Which operating system(s) are you seeing the problem on?

Linux, In deployment

### Which CPU architectures are you seeing the problem on?

x86_64, In deployment

### Relevant log output

_No response_

### Duplicate declaration

- [X] I have searched the issues and there are none like this.

### PR DESCRIPTION
**Pull request by jonaro00**
## Description of change

- Executables are moved to `{project_root}/.shuttle-executables/{uuid}` to simplify the storage environment. Binaries are executed from inside the workspace root. Closes #1008, Closes https://github.com/shuttle-hq/shuttle-docs/issues/174
- Static assets can be declared like this, to include arbitrary files in your upload. Closes #808
  ```toml
  name = "blballba"
  # this list is for including any files that was .gitignore'd or .ignore'd.
  # future plans: the files declared here + your built executable are moved from builder to runtime container
  assets = [
    "file.txt",
    "frontend/dist/*",
    "static/*",
  ]
  ```
  - Deprecated static-folder. Just returns the PathBuf for now.
- Build time env vars should now be possible with `[env]` in `.cargo/config.toml`. Closes #928 and partially closes ( #703 )

## How has this been tested? (if applicable)

Have tested locally a bunch.

### CODE REVIEW

**Review by chesedo — COMMENTED** [review]
Looking good! I left a few comments

File: common/src/deployment.rs [hunk 1]
```diff
@@ -21,13 +23,33 @@ pub enum State {
     Unknown,
 }
 
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct DeploymentMetadata {
+    pub env: Environment,
+    pub service_name: String, // TODO: replace with proper string wrapper, sort out their difference
```
[the remark this answers is not in the document]
Comment (jonaro00) [reply common/src/deployment.rs:29]: Hmm, yeah 😞

File: deployer/prepare.sh [hunk 2]
```diff
@@ -39,8 +39,10 @@ if [[ $PROD != "true" ]]; then
     shuttle-warp = { path = "/usr/src/shuttle/services/shuttle-warp" }' > $CARGO_HOME/config.toml
 fi
 
-# Add the wasm32-wasi target
+# Add the wasm32-wasi target for next
 rustup target add wasm32-wasi
+# Add the wasm32 target for frontend frameworks
+rustup target add wasm32-unknown-unknown
```
[the remark this answers is not in the document]
Comment (jonaro00) [reply deployer/prepare.sh:45]: Sorry, was squeezing in semi-support for Trunk & building your Rust frontend (someone in community does this already). 😬 
Hope it's okay.

File: common/src/models/deployment.rs [hunk 3]
```diff
@@ -13,6 +13,20 @@ use std::{fmt::Display, str::FromStr};
...
+pub const CREATE_SERVICE_BODY_LIMIT: usize = 50_000_000;
+const GIT_OPTION_NONE_TEXT: &str = "N/A";
+
+//
+// Constants regarding the deployer environment and conventions
+//
+/// Where executables are moved to in order to persist across deploys, relative to workspace root
+pub const EXECUTABLE_DIRNAME: &str = ".shuttle-executables";
+/// Where general files will persist across deploys, relative to workspace root. Used by plugins.
+pub const STORAGE_DIRNAME: &str = ".shuttle-storage";
```
[the remark this answers is not in the document]
Comment (jonaro00) [reply common/src/models/deployment.rs:28]: Well, they are not models, but they are deployment related constants. 🤔

File: deployer/prepare.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (chesedo) [reply deployer/prepare.sh:45]: I understand why our community wants this. But it should be in a separate PR

**Comment by jonaro00** [comment]
@chesedo Do we want to include a migration here that removes static folder resources, or leave that for later? (IMO leave it for later)

**Comment by jonaro00** [comment]
@chesedo One big implication of this change is that projects that write files to disk (such as rustypaste) will write them in the volume instead of the image, although they will be cleared on a new deployment if written inside the workspace folder (workdir).
Also... deploying something will instantly update static content that is being served, before the new binary has been built. More motivation to finish the builder I guess... If this is a big issue, we can find a workaround.

File: deployer/src/deployment/queue.rs [hunk 4]
```diff
@@ -338,7 +359,8 @@ async fn build_deployment(
     project_path: &Path,
     tx: crossbeam_channel::Sender<Message>,
 ) -> Result<BuiltService> {
-    let runtimes = build_workspace(project_path, true, tx, true)
+    let release = std::env::var("CI").is_ok_and(|s| s == "true");
```
Comment (jonaro00) [inline deployer/src/deployment/queue.rs:362]: Double check all environments and comment

File: resources/turso/src/lib.rs [hunk 5]
```diff
@@ -99,21 +102,20 @@ impl ResourceBuilder<Client> for Turso {
...
                     None => {
-                        let conn_url = format!(
-                            "file://{}",
-                            default_db_path
-                                .to_str()
-                                .expect("local db should be a valid unicode string")
-                        );
+                        // Default to a local db of the name of the service.
+                        let db_file = std::env::current_dir() // Should be root of the project's workspace
+                            .and_then(std::fs::canonicalize)
```
Comment (oddgrd) [inline resources/turso/src/lib.rs:111]: Might be worth testing this on Windows to see if we need dunce.

**Review by oddgrd — APPROVED** [review]
Thanks, this looks great!

**Review by iulianbarbu — APPROVED** [review]

**Pull request merged**